File: backend/src/services/sessionStateService.js
import { SessionIntent, SessionStatus } from "../constants.js";

export function initializeSessionState(chatSession) {
  let stage = "PROFILE_CAPTURED";
  let status = SessionStatus.PROFILE_CAPTURED;
  if (chatSession.intent === SessionIntent.UPLOAD_DOCUMENT) {
    stage = "AWAITING_DOCUMENT_UPLOAD";
    status = SessionStatus.AWAITING_DOCUMENT;
  }

  const snapshot = {
    stage,
    intent: chatSession.intent,
    profile_complete: true,
    document_uploaded: false,
    last_user_message: null,
    last_assistant_message: null,
    last_tool_names: [],
    updated_at: new Date().toISOString(),
  };
  chatSession.status = status;
  chatSession.state_snapshot = snapshot;
  return snapshot;
}

export function advanceStateAfterMessage({
  chatSession,
  userMessage,
  assistantMessage,
  toolOutputs,
}) {
  const snapshot = {
    ...(chatSession.state_snapshot || {}),
    stage: "ADVISORY_IN_PROGRESS",
    intent: chatSession.intent,
    last_user_message: userMessage,
    last_assistant_message: assistantMessage,
    last_tool_names: Object.keys(toolOutputs).sort(),
    updated_at: new Date().toISOString(),
  };

  if (
    chatSession.intent === SessionIntent.UPLOAD_DOCUMENT &&
    !snapshot.document_uploaded
  ) {
    chatSession.status = SessionStatus.AWAITING_DOCUMENT;
  } else {
    chatSession.status = SessionStatus.ACTIVE;
  }

  chatSession.state_snapshot = snapshot;
  return snapshot;
}

export function advanceStateAfterDocument({ chatSession, filename }) {
  const snapshot = {
    ...(chatSession.state_snapshot || {}),
    stage: "DOCUMENT_PROCESSED",
    document_uploaded: true,
    document_name: filename,
    updated_at: new Date().toISOString(),
  };
  chatSession.status = SessionStatus.ACTIVE;
  chatSession.state_snapshot = snapshot;
  return snapshot;
}

export function advanceStateAfterSummary(chatSession) {
  const snapshot = {
    ...(chatSession.state_snapshot || {}),
    stage: "SUMMARY_READY",
    summary_ready: true,
    updated_at: new Date().toISOString(),
  };
  chatSession.status = SessionStatus.READY_FOR_SUMMARY;
  chatSession.state_snapshot = snapshot;
  return snapshot;
}

File: backend/app/services/session_state_service.py
from datetime import datetime
from typing import Any

from app.models.chat import ChatSession, SessionIntent, SessionStatus


def initialize_session_state(chat_session: ChatSession) -> dict[str, Any]:
    stage = "PROFILE_CAPTURED"
    status = SessionStatus.PROFILE_CAPTURED
    if chat_session.intent == SessionIntent.UPLOAD_DOCUMENT:
        stage = "AWAITING_DOCUMENT_UPLOAD"
        status = SessionStatus.AWAITING_DOCUMENT

    snapshot = {
        "stage": stage,
        "intent": chat_session.intent.value,
        "profile_complete": True,
        "document_uploaded": False,
        "last_user_message": None,
        "last_assistant_message": None,
        "last_tool_names": [],
        "updated_at": datetime.utcnow().isoformat(),
    }
    chat_session.status = status
    chat_session.state_snapshot = snapshot
    return snapshot


def advance_state_after_message(
    *,
    chat_session: ChatSession,
    user_message: str,
    assistant_message: str,
    tool_outputs: dict[str, Any],
) -> dict[str, Any]:
    snapshot = dict(chat_session.state_snapshot or {})
    snapshot.update(
        {
            "stage": "ADVISORY_IN_PROGRESS",
            "intent": chat_session.intent.value,
            "last_user_message": user_message,
            "last_assistant_message": assistant_message,
            "last_tool_names": sorted(tool_outputs.keys()),
            "updated_at": datetime.utcnow().isoformat(),
        }
    )

    if chat_session.intent == SessionIntent.UPLOAD_DOCUMENT and not snapshot.get("document_uploaded"):
        chat_session.status = SessionStatus.AWAITING_DOCUMENT
    else:
        chat_session.status = SessionStatus.ACTIVE

    chat_session.state_snapshot = snapshot
    return snapshot


def advance_state_after_document(
    *,
    chat_session: ChatSession,
    filename: str,
) -> dict[str, Any]:
    snapshot = dict(chat_session.state_snapshot or {})
    snapshot.update(
        {
            "stage": "DOCUMENT_PROCESSED",
            "document_uploaded": True,
            "document_name": filename,
            "updated_at": datetime.utcnow().isoformat(),
        }
    )
    chat_session.status = SessionStatus.ACTIVE
    chat_session.state_snapshot = snapshot
    return snapshot


def advance_state_after_summary(chat_session: ChatSession) -> dict[str, Any]:
    snapshot = dict(chat_session.state_snapshot or {})
    snapshot.update(
        {
            "stage": "SUMMARY_READY",
            "summary_ready": True,
            "updated_at": datetime.utcnow().isoformat(),
        }
    )
    chat_session.status = SessionStatus.READY_FOR_SUMMARY
    chat_session.state_snapshot = snapshot
    return snapshot
